File: back/persistence/model/src/main/kotlin/Notification.kt
@file:OptIn(ExperimentalTime::class)

package persistence.model

import id.Id
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.time.ExperimentalTime
import kotlin.time.Instant

/**
 * A user-facing notification, materialised as a first-class synced entity on the
 * recipient's private scope (see ADR-005).
 *
 * Notifications are server-authoritative: a domain mutation's side-effect creates the
 * row and its scope [persistence.changes.Change] atomically. The client only ever
 * mutates [readAt] (mark read) or deletes the row (archive) through `POST /v1/sync`.
 *
 * [recipientScope] is the private scope key the notification belongs to —
 * `member:{sub}`, `owner:{sub}`, and `producer-account:{id}` are all active now.
 * A single [Notification] type serves every recipient kind; the difference is carried
 * by [recipientScope] + [category].
 */
@Serializable
data class Notification(
    @SerialName("notification_id") val notificationId: Id<Notification>,
    @SerialName("recipient_scope") val recipientScope: String,
    val type: NotificationType,
    val category: NotificationCategory,
    val title: String,
    val body: String,
    @SerialName("deep_link") val deepLink: String? = null,
    @SerialName("related_entity_id") val relatedEntityId: String? = null,
    @SerialName("created_at") val createdAt: Instant,
    @SerialName("read_at") val readAt: Instant? = null,
)

/** Severity / intent of a [Notification]. Drives client styling and ordering. */
@Serializable
enum class NotificationType {
    ALERT,
    REMINDER,
    INFO,
    URGENT,
}

/**
 * Business origin of a [Notification]. Lets the client group/filter and lets the
 * transport layer pick a per-category template. Extend as new event producers are wired.
 */
@Serializable
enum class NotificationCategory {
    GENERIC,
    BASKET_EXCHANGE_REQUEST_RECEIVED,
    BASKET_EXCHANGE_ACCEPTED,
    BASKET_EXCHANGE_REJECTED,
    ORGANIZATION_REQUEST_SUBMITTED,
    PRODUCER_REQUEST_SUBMITTED,
    MEMBER_JOIN_REQUEST_SUBMITTED,
    DELIVERY_REMINDER,
    SLOT_CANCELLED,
    SLOT_RESCHEDULED,
}

/**
 * Outbound transport channel for a notification. The in-app feed always exists (it is
 * the synced [Notification] itself); these are the *additional* push transports a
 * recipient may opt into via their preferences.
 */
@Serializable
enum class NotificationChannel {
    EMAIL,
    PUSH,
}

/**
 * Admin-authored override of the title/body used for a given [NotificationCategory]
 * within an organization (see `Organization.notificationOverrides`). Either field may be
 * null/blank, in which case the hardcoded default copy is used for that part. The override
 * is applied verbatim — no variable interpolation.
 */
@Serializable
data class NotificationCopyOverride(
    val title: String? = null,
    val body: String? = null,
)

File: front/lib/domain/model/notification.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'notification.freezed.dart';
part 'notification.g.dart';

/// A user-facing notification on the recipient's private sync feed (ADR-005).
///
/// Mirrors `Notification` in
/// `back/persistence/model/src/main/kotlin/Notification.kt`. Server-authoritative:
/// the client only ever flips [readAt] (mark read) or deletes the row (archive).
/// [createdAt] / [readAt] are ISO-8601 instants on the wire.
@freezed
abstract class AppNotification with _$AppNotification {
  const factory AppNotification({
    @JsonKey(name: 'notification_id') required String notificationId,
    @JsonKey(name: 'recipient_scope') required String recipientScope,
    required NotificationType type,
    required NotificationCategory category,
    required String title,
    required String body,
    @JsonKey(name: 'deep_link') String? deepLink,
    @JsonKey(name: 'related_entity_id') String? relatedEntityId,
    @JsonKey(name: 'created_at') required String createdAt,
    @JsonKey(name: 'read_at') String? readAt,
  }) = _AppNotification;

  factory AppNotification.fromJson(Map<String, Object?> json) =>
      _$AppNotificationFromJson(json);
}

/// Severity / intent. Wire values uppercase, matching the back enum.
enum NotificationType {
  @JsonValue('ALERT')
  alert,
  @JsonValue('REMINDER')
  reminder,
  @JsonValue('INFO')
  info,
  @JsonValue('URGENT')
  urgent,
}

/// Business origin of the notification. Wire values uppercase, matching the back enum.
enum NotificationCategory {
  @JsonValue('GENERIC')
  generic,
  @JsonValue('BASKET_EXCHANGE_REQUEST_RECEIVED')
  basketExchangeRequestReceived,
  @JsonValue('BASKET_EXCHANGE_ACCEPTED')
  basketExchangeAccepted,
  @JsonValue('BASKET_EXCHANGE_REJECTED')
  basketExchangeRejected,
  @JsonValue('MEMBER_JOIN_REQUEST_SUBMITTED')
  memberJoinRequestSubmitted,
  @JsonValue('DELIVERY_REMINDER')
  deliveryReminder,
  @JsonValue('ORGANIZATION_REQUEST_SUBMITTED')
  organizationRequestSubmitted,
  @JsonValue('PRODUCER_REQUEST_SUBMITTED')
  producerRequestSubmitted,
  @JsonValue('SLOT_CANCELLED')
  slotCancelled,
  @JsonValue('SLOT_RESCHEDULED')
  slotRescheduled,
}
